src/utils/index.ts

```diff
@@ -15,11 +15,50 @@ export class TimeoutError extends Error {
     }
 }
 
+
+export interface ICancellable {
+    cancel(): void;
+}
+
+/**
+ * A promise that can be canceled to release any resource.
+ * Instances of this class should guarantee that all resources will eventually be released if `cancel()` is called,
+ * regardless of wether the Promise is fulfilled or rejected.
+ *
+ * Once `cancel()` is called, the behaviour of the promise is undefined, and the caller
+ * should not expect it to reject or fulfill, nor to be pending forever.
+ */
+export class CancellablePromise<T> extends Promise<T> implements ICancellable {
+    private mCancelled = false;
+
+    public get cancelled(): boolean {
+        return this.mCancelled;
+    }
+
+    constructor(
+        executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void,
+        private canceller?: () => void
```

Comment:
The reason I put it optional because waitFor above was not using an explicit cancellor (relying on the cancelled member). I'll change it, as I'm removing waitFor.